faq.md

```diff
@@ -499,6 +499,90 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
+[Iterable](https://docs.flutter.io/flutter/dart-core/Iterable-class.html) API,
+which makes heavy use of the functional style (map, reduce, where, etc), is
+frequently used to process lists of values in the framework.
+
+* Event-driven programming: User interactions are represented by event objects that are
+dispatched to callbacks registered with event handlers. Screen updates are triggered
+by a similar callback mechanism. The [Listenable](https://docs.flutter.io/flutter/foundation/Listenable-class.html)
+class, which is used as the basis of the animation system, formalises a
+subscription model for events with multiple listeners.
+
+* Class-based object-orientated programming: Most of the APIs of the framework are built using classes
+with inheritance. We use an approach whereby we define very high-level APIs in our base classes, then
+specialise them iteratively in subclasses. For example, our render objects have a base class
+([RenderObject](https://docs.flutter.io/flutter/rendering/RenderObject-class.html))
+that is agnostic regarding the coordinate system, and then we have a subclass
+([RenderBox](https://docs.flutter.io/flutter/rendering/RenderBox-class.html))
+that introduces the opinion that the geometry should be based on the Cartesian coordinate system (x/width and y/height).
+
+* Prototype-based object-orientated programming: The
+[ScrollPhysics](https://docs.flutter.io/flutter/widgets/ScrollPhysics-class.html) class
+chains instances to compose the physics that apply to scrolling dynamically at runtime. This
+lets the system compose, for example, paging physics with platform-specific physics, without
+the platform having to be selected at compile time.
+
+* Imperative programming: Straight forward imperative programming, usually paired with state encapsulated
+within an object, is used where it provides the most intuitive solution. For example, tests are
+written in an imperative style, first describing the situation under test, then listing the invariants
+that the test must match, then advancing the clock or inserting events as necessary for the test.
+
+* Reactive programming: The widget and element trees are sometimes described as reactive, because new inputs 
+provided in a widget's constructor are immediately propagated as changes to lower-level widgets by
+the widget's build method, and changes made in the lower widgets (e.g. in response to user input) propagate
+back up the tree via event handlers. Aspects of both functional-reactive and imperative-reactive are
+present in the framework. 
+
+* Declarative programming: The build functions of widgets are typically a single expression with multiple levels
+of nested constructors, written using a strictly declarative subset of Dart (such nested expressions could be mechanically
+transformed to or from any suitably expressive markup language). This can also be combined with
+the imperative style to build UIs that would be harder to describe in a pure-declarative approach.
```

Comment:
Definitely need an example, and a statement of the benefit or problem solved. Does this allow Flutter to reorder the building of the widget tree, or do transformations on the code prior to building the tree? If not, then this is imperative since it also proscribes how the tree must be created, not just what it looks like. Basically, why did you build the tree this way? What's the advantage?

Reply:
> Definitely need an example, and a statement of the benefit or problem solved.

As above, I think we should avoid getting too specific because that would take away from the point of the FAQ entry. We give a pointer for people who want to learn more ("build methods").

> Does this allow Flutter to reorder the building of the widget tree, or do transformations on the code prior to building the tree? If not, then this is imperative since it also proscribes how the tree must be created, not just what it looks like.

Flutter does internally control the order in which widgets are built, but I don't know that one could really say that the declarative style "allows" it. Just being declarative doesn't necessarily mean that things can be arbitrarily reordered; if anything, it is usually used to _force_ an order (by requiring that the system determine the required order of operations).